llama_hub/tools/neo4j_db/base.py

```diff
@@ -0,0 +1,146 @@
+            result = session.run(neo4j_query, params)
+            output = [r.values() for r in result]
+            output.insert(0, list(result.keys()))
+            return output
+
+    def query_graph_db_as_document(self, neo4j_query, params=None):
+        """
+        Queries the Neo4j database.
+
+        Args:
+            neo4j_query (str): The Cypher query to be executed.
+            params (dict, optional): Parameters for the Cypher query. Defaults to None.
+
+        Returns:
+            list: The query results.
+        """
+        if params is None:
+            params = {}
+        with self.graph_store.client.session() as session:
+            result = session.run(neo4j_query, params)
+            output = [Document(text="\n".join([f"{key}:{val}" for key, val in zip(r.keys(), r.values())])) for r in result]
+            return output
+
+    def construct_cypher_query(self, question, history=None):
+        """
+        Constructs a Cypher query based on a given question and history.
+
+        Args:
+            question (str): The question to construct the Cypher query for.
+            history (list, optional): A list of previous interactions for context. Defaults to None.
+
+        Returns:
+            str: The constructed Cypher query.
+        """
+        messages = [
+            ChatMessage(role='system', content=self.get_system_message()),
+            ChatMessage(role='user', content=question),
+        ]
+        # Used for Cypher healing flows
+        if history:
+            messages.extend(history)
+
+        completions = self.llm.chat(messages)
+        return completions.message.content
+
+    def run(self, question, history=None, retry=True):
+        """
+        Executes a Cypher query based on a given question.
+
+        Args:
+            question (str): The question to execute the Cypher query for.
+            history (list, optional): A list of previous interactions for context. Defaults to None.
+            retry (bool, optional): Whether to retry in case of a syntax error. Defaults to True.
+
+        Returns:
+            list/str: The query results or an error message.
+        """
+        from neo4j.exceptions import CypherSyntaxError
+
+        # Construct Cypher statement
+        cypher = self.construct_cypher_query(question, history)
+        print(cypher)
+        try:
+            return self.query_graph_db_as_document(cypher)
+        # Self-healing flow
+        except CypherSyntaxError as e:
+            # If out of retries
+            if not retry:
+                return "Invalid Cypher syntax"
+            # Self-healing Cypher flow by
+            # providing specific error to GPT-4
+            print("Retrying")
+            return self.run(
+                question,
+                [
+                    ChatMessage(role='assistant', content=cypher),
+                    ChatMessage(role='system', conent=f"""This query returns an error: {str(e)} 
```

Comment:
nit: due to the triple quotes you're going to get a query string with a lot of spaces between "This query returns an error: {str(e)}" and "Give me a...". Try inserting a backslash after the first line to not insert a space.